paddle/fluid/framework/fleet/box_wrapper.h

```diff
@@ -46,44 +136,185 @@ class BoxWrapper {
                       const std::vector<const uint64_t*>& keys,
                       const std::vector<const float*>& grad_values,
                       const std::vector<int64_t>& slot_lengths,
-                      const int hidden_size);
+                      const int hidden_size, const int batch_size);
+  void CopyForPull(const paddle::platform::Place& place, uint64_t** gpu_keys,
+                   const std::vector<float*>& values,
+                   const boxps::FeatureValueGpu* total_values_gpu,
+                   const int64_t* gpu_len, const int slot_num,
+                   const int hidden_size, const int64_t total_length);
+  void CopyForPush(const paddle::platform::Place& place,
+                   const std::vector<const float*>& grad_values,
+                   boxps::FeaturePushValueGpu* total_grad_values_gpu,
+                   const std::vector<int64_t>& slot_lengths,
+                   const int hidden_size, const int64_t total_length,
+                   const int batch_size);
+  void CopyKeys(const paddle::platform::Place& place, uint64_t** origin_keys,
+                uint64_t* total_keys, const int64_t* gpu_len, int slot_num,
+                int total_len);
+  boxps::PSAgentBase* GetAgent() { return p_agent_; }
+  void InitializeGPU(const char* conf_file, const std::vector<int>& slot_vector,
+                     const std::vector<std::string>& slot_omit_in_feedpass) {
+    if (nullptr != s_instance_) {
+      VLOG(3) << "Begin InitializeGPU";
+      std::vector<cudaStream_t*> stream_list;
+      for (int i = 0; i < platform::GetCUDADeviceCount(); ++i) {
+        VLOG(3) << "before get context i[" << i << "]";
+        platform::CUDADeviceContext* context =
+            dynamic_cast<platform::CUDADeviceContext*>(
+                platform::DeviceContextPool::Instance().Get(
+                    platform::CUDAPlace(i)));
+        stream_list_[i] = context->stream();
+        stream_list.push_back(&stream_list_[i]);
+      }
+      VLOG(2) << "Begin call InitializeGPU in BoxPS";
+      // the second parameter is useless
+      s_instance_->boxps_ptr_->InitializeGPU(conf_file, -1, stream_list);
+      p_agent_ = boxps::PSAgentBase::GetIns(feedpass_thread_num_);
+      p_agent_->Init();
+      for (const auto& slot_name : slot_omit_in_feedpass) {
+        slot_name_omited_in_feedpass_.insert(slot_name);
+      }
+      slot_vector_ = slot_vector;
+      keys_tensor.resize(platform::GetCUDADeviceCount());
+    }
+  }
+
+  int GetFeedpassThreadNum() const { return feedpass_thread_num_; }
+
+  void Finalize() {
+    VLOG(3) << "Begin Finalize";
+    if (nullptr != s_instance_) {
+      s_instance_->boxps_ptr_->Finalize();
+    }
+  }
+
+  void SaveBase(const char* batch_model_path, const char* xbox_model_path,
+                boxps::SaveModelStat& stat) {  // NOLINT
```

Comment:
it is not good to expose `boxps` namespace out of box_wrapper, it is better to wrap `boxps`-related data in a paddle-friendly data structure.

Reply:
Thanks, will make it a future work